### PR TITLE
Enable training with fastText indexer.

### PR DESCRIPTION
**Pull request by sebpuetz**
Add support to train with the fastText hashing function instead
of fnv1a.

### CODE REVIEW

File: src/subcommands/traits.rs [hunk 1]
```diff
@@ -233,12 +245,20 @@ where
                     .map(|v| v.parse().context("Cannot parse bucket exponent"))
                     .transpose()?
                     .unwrap();
+                let indexer = matches
+                    .value_of(HASH_INDEXER_TYPE)
+                    .map(|v| v.try_into().context("Cannot subword indexer type"))
```
Comment (danieldk) [inline src/subcommands/traits.rs:250]: Missing verb?

Comment (sebpuetz) [reply src/subcommands/traits.rs]: oops

**Comment by sebpuetz** [comment]
> Did you have any chance yet to train some embeddings to test?

I ran a couple of toy-models on tüba-dz to verify that there's no unexpected panics etc. 

------

I made some more semantic changes, e.g. the buckets value was called `bucket_exp` which doesn't fit the fastText indexer, the description of the `buckets` CLI option didn't reflect that either.

**Comment by danieldk** [comment]
> > Did you have any chance yet to train some embeddings to test?
> 
> I ran a couple of toy-models on tüba-dz to verify that there's no unexpected panics etc.
> 
> I made some more semantic changes, e.g. the buckets value was called `bucket_exp` which doesn't fit the fastText indexer, the description of the `buckets` CLI option didn't reflect that either.

I think I'd prefer keeping `bucket_exp` and setting the number of fastText buckets to 2^n then. It's the more principled way, since it does not introduce a bias for certain buckets.

**Comment by sebpuetz** [comment]
> > > Did you have any chance yet to train some embeddings to test?
> > 
> > 
> > I ran a couple of toy-models on tüba-dz to verify that there's no unexpected panics etc.
> > I made some more semantic changes, e.g. the buckets value was called `bucket_exp` which doesn't fit the fastText indexer, the description of the `buckets` CLI option didn't reflect that either.
> 
> I think I'd prefer keeping `bucket_exp` and setting the number of fastText buckets to 2^n then. It's the more principled way, since it does not introduce a bias for certain buckets.

Reverted the doc changes etc, `From<VocabBuilder>` for the subword vocabs contains a bit of a hack now since the indexers need to be handled differently.

```Rust
let buckets = match config.indexer.indexer_type {
    BucketIndexerType::Finalfusion => config.indexer.buckets_exp as usize,
    BucketIndexerType::FastText => 2u64.pow(config.indexer.buckets_exp) as usize,
};
SubwordVocab::new(config, words, builder.n_items, I::new(buckets))
```

**Comment by sebpuetz** [comment]
Is this good to go with the reverted changes?

**Pull request merged**